subprojects/publish/src/main/java/org/gradle/api/publish/internal/mapping/DefaultDependencyCoordinateResolverFactory.java

```diff
@@ -41,33 +46,79 @@
     }
 
     @Override
     public DependencyResolvers createCoordinateResolvers(SoftwareComponentVariant variant, VersionMappingStrategyInternal versionMappingStrategy) {
         Configuration configuration = null;
+        if (variant instanceof ResolutionBackedVariant) {
+            ResolutionBackedVariant resolutionBackedVariant = (ResolutionBackedVariant) variant;
+            configuration = resolutionBackedVariant.getResolutionConfiguration();
+
+            boolean useResolvedCoordinates = resolutionBackedVariant.getPublishResolvedCoordinates();
+            if (useResolvedCoordinates && configuration == null) {
+                throw new InvalidUserDataException("Cannot enable dependency mapping without configuring a resolution configuration.");
+            } else if (useResolvedCoordinates) {
+
+                ComponentDependencyResolver componentResolver = new ResolutionBackedComponentDependencyResolver(
+                    configuration,
+                    moduleIdentifierFactory,
+                    projectDependencyResolver
+                );
+
+                VariantDependencyResolver variantResolver = new ResolutionBackedVariantDependencyResolver(
+                    projectDependencyResolver,
+                    moduleIdentifierFactory,
+                    configuration,
+                    attributeDesugaring,
+                    componentResolver
+                );
+
+                return new DependencyResolvers(variantResolver, componentResolver);
+            }
+        }
 
         ImmutableAttributes attributes = ((AttributeContainerInternal) variant.getAttributes()).asImmutable();
         VariantVersionMappingStrategyInternal versionMapping = versionMappingStrategy.findStrategyForVariant(attributes);
 
         // Fallback to component coordinate mapping if variant mapping is not enabled
-        ComponentDependencyResolver componentResolver;
+        ComponentDependencyResolver componentResolver = null;
         if (versionMapping.isEnabled()) {
             if (versionMapping.getUserResolutionConfiguration() != null) {
                 configuration = versionMapping.getUserResolutionConfiguration();
-            } else if (versionMapping.getDefaultResolutionConfiguration() != null) {
+            } else if (versionMapping.getDefaultResolutionConfiguration() != null && configuration == null) {
+                // The configuration set on the variant is almost always more correct than the
+                // default version mapping configuration, which is currently set project-wide
+                // by the Java plugin. For this reason, we only use the version mapping default
+                // if the dependency mapping configuration is not set.
                 configuration = versionMapping.getDefaultResolutionConfiguration();
             }
+
+            if (configuration != null) {
+                componentResolver = USE_LEGACY_VERSION_MAPPING
```

Comment:
❗ This is assignment, it's always `true`.

Reply:
Why didn't this fail a test?

Reply:
Yes, it is effectively an assignment, however `USE_LEGACY_VERSION_MAPPING` has a comment saying we should switch it to false when we are comfortable with the new functionality. 
This should change relatively soon after we add more extensive testing. 

I am also considering mapping `USE_LEGACY_VERSION_MAPPING` to a system property so that we can run all the version mapping tests against the new functionality.

Reply:
Can you make the comment on `USE_LEGACY_VERSION_MAPPING` into javadoc then, so it shows up on mouseover?  I'd also consider reworking this to a full `if-else` vs. the ternary - it also looks like line 114 is a typo, since the types don't match.